# Practica aprenentage automatic

    Juan José Puya Milián  DNI: 43229502J
    Enlace repositorio github: https://github.com/juanjopuya/Practica_IA_Juanjo_Puya


## Introducción


Tenemos a nuestra disposición un conjunto de datos sobre las caracteristicas de todos los jugadores que se encuentran en el videojuego **FIFA**. El objectivo de esta práctica es realizar una **limpieza de nuestro dataset** para poder entrenar nuestro algoritmo de machine learning, y de esta forma predecir el **valor** de un jugador cualquiera con datos los datos que tenemos en la base de datos


### Importar las librerías

In [1550]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

### Lectura de datos

In [1551]:
df = pd.read_csv(os.path.join("in", "fifa.csv"))
pd.set_option('display.max_columns', None)
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

### Analisis de los datos

Uno de los primeros pasos a realizar es la eliminacion de todos esos datos que no son relevantes para nuestro entrenamiento. Algunas de las columnas que se van a eliminar se puede observar a simple vista que no sirven, como por ejemplo las urls de los png. Otras se han eliminado después de valorar sobre cada un de ellas si de verdad aportan información util o no. Por ejemplo, ``Real Face`` no aporta nada a nuestro entrenamiento porque que si tiene o no la cara hecha no va a influir en nada sobre su valor. Otros casos similares son el ``Jersey Number`` o ``Body Type``. Por otra parte tenemos casos como ``Joined``,``Loaned From`` o ``Contract Valid Until`` que los he eliminado porque he considerado que no son suficientemente importantes para manternelos en el dataset ya que tampoco tienen relacion con el posible valor de cada jugador.

La columna ``Name`` no se ha eliminado de momento para tener de forma visual el nombre de los jugadores y así saber de que jugador estamos tratando los datos en cada momento. Esta columna se eliminará al final ya que no la necesitaremos para este entrenamiento.

In [1552]:
df.drop(['Unnamed: 0','ID','Photo','Flag','Club Logo','Real Face','Body Type','Contract Valid Until','Work Rate','Jersey Number','Joined','Loaned From'],axis=1,inplace=True)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club    Value   Wage  Special Preferred Foot  \
0         FC Barcelona  €110.5M  €565K     2202           Left   
1             Juventus     €77M  €405K     2228          Right   
2  Paris Saint-Germain  €118.5M  €290K     2143          Right   
3    Manchester United     €72M  €260K     1471          Right   
4      Manchester City    €102M  €355K     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position Height  Weight  \
0                       5.0        4.0          4.0       RF    5'7  159lbs   
1                       5.0        4.0          5.0       ST    6'2  183lbs   
2                       5.0        5.0          5.0       LW    5'9  150lbs   
3                       4.0        3.0          1.0       GK    6'4  168lbs   
4                       4.0        5.0          4.0      RCM   5'11  154lbs   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Mark

## Tratamiento de los Nans

Los valores nulos o ‘NaN’ indican que no hay registro. Este tipo de valores no pueden estar en el momento de hacer el entrenamiento.

Con la siguiente instrucción obtenemos una lista de todas las columnas de nuestra tabla que contienen NaNs.

In [1553]:
df.columns[df.isna().any()].tolist()

['Club',
 'Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Position',
 'Height',
 'Weight',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

Como se puede observar, la gran mayoría de las columnas contienen NaNs, ante esto tenemos varias opciones: deducir el valor del registro de otra información, completarlo con otro valor o eliminarlo. 

Vamos a analizar la columna ``Club``. Podemos obtener todos los jugadores que no está especificado en que equipo juegan, y después valorar como procedemos con el tratamiento de los NaNs.


In [1554]:
df.loc[df.Club != df.Club]

Name  Age Nationality  Overall  Potential Club Value Wage  \
452        L. Paredes   24   Argentina       80         85  NaN    €0   €0   
538      A. Granqvist   33      Sweden       80         80  NaN    €0   €0   
568          A. Lunev   26      Russia       79         81  NaN    €0   €0   
677      I. Smolnikov   29      Russia       79         79  NaN    €0   €0   
874         A. Dzyuba   29      Russia       78         78  NaN    €0   €0   
...               ...  ...         ...      ...        ...  ...   ...  ...   
17197        V. Kaith   21       India       55         64  NaN    €0   €0   
17215        K. Taimi   26     Finland       55         57  NaN    €0   €0   
17339        S. Passi   23       India       54         63  NaN    €0   €0   
17436  D. Lalhlimpuia   20       India       54         67  NaN    €0   €0   
17539        C. Singh   21       India       53         62  NaN    €0   €0   

       Special Preferred Foot  International Reputation  Weak Foot  \
452       2122          Right                       2.0        4.0   
538       1797          Right                       2.0        4.0   
568       1217          Right                       1.0        3.0   
677       2038          Right                       2.0        3.0   
874       1810          Right                       2.0        3.0   
...        ...            ...                       ...        ...   
17197      838          Right                       1.0        2.0   
17215     1366          Right                       1.0        3.0   
17339     1321          Right                       1.0        3.0   
17436     1270          Right                       1.0        3.0   
17539     1247          Right                       1.0        3.0   

       Skill Moves Position Height  Weight    LS    ST    RS    LW    LF  \
452            4.0       CM   5'11  165lbs  71+2  71+2  71+2  75+2  75+2   
538            2.0      LCB    6'4  185lbs  62+2  62+2  62+2  56+2  58+2   
568            1.0       GK    6'2  176lbs   NaN   NaN   NaN   NaN   NaN   
677            3.0       RB   5'10  154lbs  70+2  70+2  70+2  73+2  72+2   
874            3.0       ST    6'5  201lbs  77+2  77+2  77+2  71+2  74+2   
...            ...      ...    ...     ...   ...   ...   ...   ...   ...   
17197          1.0       GK    6'2  176lbs   NaN   NaN   NaN   NaN   NaN   
17215          2.0       RB    6'4  187lbs  46+2  46+2  46+2  45+2  44+2   
17339          2.0      NaN    5'9  143lbs   NaN   NaN   NaN   NaN   NaN   
17436          2.0      NaN    6'0  168lbs   NaN   NaN   NaN   NaN   NaN   
17539          2.0      NaN    6'3  174lbs   NaN   NaN   NaN   NaN   NaN   

         CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB  \
452    75+2  75+2  75+2  77+2  77+2  77+2  76+2  79+2  79+2  79+2  76+2  75+2   
538    58+2  58+2  56+2  58+2  58+2  58+2  57+2  64+2  64+2  64+2  57+2  68+2   
568     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
677    72+2  72+2  73+2  73+2  73+2  73+2  75+2  74+2  74+2  74+2  75+2  78+2   
874    74+2  74+2  71+2  71+2  71+2  71+2  71+2  66+2  66+2  66+2  71+2  52+2   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17197   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17215  44+2  44+2  45+2  43+2  43+2  43+2  46+2  43+2  43+2  43+2  46+2  51+2   
17339   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17436   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17539   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
452    77+2  77+2  77+2  75+2  74+2  72+2  72+2  72+2  74+2      76.0   
538    74+2  74+2  74+2  68+2  70+2  79+2  79+2  79+2  70+2      49.0   
568     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      16.0   
677    75+2  75+2  75+2  78+2  78+2  73+2  73+2  73+2  78+2  

Tenemos una cantidad bastante pequeña de jugadores sin club especificado en comparación con todo nuestro dataset, por lo que una opción válida será eliminar todos esos jugadores ya que nos presentan un problema.

In [1555]:
df = df.dropna(subset = ['Club'])
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club    Value   Wage  Special Preferred Foot  \
0         FC Barcelona  €110.5M  €565K     2202           Left   
1             Juventus     €77M  €405K     2228          Right   
2  Paris Saint-Germain  €118.5M  €290K     2143          Right   
3    Manchester United     €72M  €260K     1471          Right   
4      Manchester City    €102M  €355K     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position Height  Weight  \
0                       5.0        4.0          4.0       RF    5'7  159lbs   
1                       5.0        4.0          5.0       ST    6'2  183lbs   
2                       5.0        5.0          5.0       LW    5'9  150lbs   
3                       4.0        3.0          1.0       GK    6'4  168lbs   
4                       4.0        5.0          4.0      RCM   5'11  154lbs   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Mark

Ahora vamos a realizar el mismo analisis con la culumna ``Position``. Primero obtenemos la lista de jugadores sin posición especificada.

In [1556]:
df.loc[df.Position != df.Position]

Name  Age       Nationality  Overall  Potential  \
13236      J. McNulty   33          Scotland       62         62   
13237      J. Barrera   29         Nicaragua       62         62   
13238        J. Stead   35           England       62         62   
13239     A. Semprini   20             Italy       62         72   
13240      R. Bingham   24           England       62         66   
13241    K. Dankowski   21            Poland       62         72   
13242       I. Colman   23         Argentina       62         70   
13243       M. Feeney   19           England       62         78   
13244        R. Minor   30           Denmark       62         62   
13245          Klauss   21            Brazil       62         69   
13246      I. Sissoko   22            France       62         68   
13247         F. Hart   28           Austria       62         62   
13248   L. McCullough   24  Northern Ireland       62         69   
13249       Li Yunqiu   27          China PR       62         62   
13250       F. Garcia   29          Paraguay       62         62   
13251    R. Haemhouts   34           Belgium       62         62   
13252       E. Binaku   22           Albania       62         70   
13253       G. Miller   31          Scotland       62         62   
13254      A. Aidonis   17           Germany       62         82   
13255        L. Sowah   25           Germany       62         65   
13256       R. Deacon   26           England       62         62   
13257   Jang Hyun Soo   25    Korea Republic       62         65   
13258     A. Al Malki   23      Saudi Arabia       62         67   
13259     E. Guerrero   27             Chile       62         65   
13260         Hernáiz   20             Spain       62         69   
13261   H. Al Mansour   25      Saudi Arabia       62         64   
13262         H. Paul   24           Germany       62         66   
13263        S. Bauer   25           Austria       62         66   
13264      M. Chergui   29            France       62         62   
13265      D. Gardner   28           England       62         62   
13266    L. Bengtsson   20            Sweden       62         73   
13267    F. Jaramillo   22          Colombia       62         70   
13268      L. Garguła   37            Poland       62         62   
13269       S. Rivera   26          Colombia       62         64   
13270        Vinicius   19            Brazil       62         77   
13271    F. Sepúlveda   26             Chile       62         63   
13272       L. Spence   22          Scotland       62         70   
13273      B. Lepistu   25           Estonia       62         67   
13274     A. Abruscia   27             Italy       62         62   
13275     E. González   23         Venezuela       62         70   
13276      M. Al Amri   26      Saudi Arabia       62         63   
13277    J. Rebolledo   26             Chile       62         62   
13278      C. Mamengi   17       Netherlands       62         79   
13279    P. Mazzocchi   22             Italy       62         69   
13280       Y. Ammour   19            France       62         77   
13281  Jwa Joon Hyeop   27    Korea Republic       62         62   
13282      O. Marrufo   25            Mexico       62         65   
13283     Han Pengfei   25          China PR       62         66   

                                Club  Value Wage  Special Preferred Foot  \
13236                       Rochdale  €120K  €1K     1510            NaN   
13237                Boyacá Chicó FC  €300K  €1K     1532            NaN   
13238                   Notts County  €140K  €3K     1573            NaN   
13239                        Brescia  €425K  €1K     1610            NaN   
13240         Hamilton Academical FC  €400K  €1K     1481            NaN   
13241                  Śląsk Wrocław  €425K  €1K     1692            NaN   
13242         Club Atlético Aldosivi  €450K  €2K     1663            NaN   
13243                        Everton  €600K  €5K     1328            NaN   


Como en el caso anterior, el porcentaje de jugadores sin posición es lo suficientemente pequeño como para eliminarlos.

In [1557]:
df = df.dropna(subset = ['Position'])
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club    Value   Wage  Special Preferred Foot  \
0         FC Barcelona  €110.5M  €565K     2202           Left   
1             Juventus     €77M  €405K     2228          Right   
2  Paris Saint-Germain  €118.5M  €290K     2143          Right   
3    Manchester United     €72M  €260K     1471          Right   
4      Manchester City    €102M  €355K     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position Height  Weight  \
0                       5.0        4.0          4.0       RF    5'7  159lbs   
1                       5.0        4.0          5.0       ST    6'2  183lbs   
2                       5.0        5.0          5.0       LW    5'9  150lbs   
3                       4.0        3.0          1.0       GK    6'4  168lbs   
4                       4.0        5.0          4.0      RCM   5'11  154lbs   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Mark

Después de este analisis, podemos sacar la cantidad de NaNs que quedan para cada columna. En esta lista se puede observar que ``Release Clause`` sigue manteniendo una cantidad de NaNs considerable. Más adalente arreglaremos este problema sustituyendo los NaNs por la media de los ``Release Clause`` restantes. Ahora no podemos hacerlo porque esta columna no está en formato númerico ya que tiene el símbolo "€" y "k". Cuando pongamos los formatos correctos se realizará este paso.

Pasa lo mismo con las variables ``LS``,``ST``,``RS``,``LW``...  Esto ocurre porque todavía no están en un formato adecuado. 

In [1558]:
df.isnull().sum()

Name                 0
Age                  0
Nationality          0
Overall              0
Potential            0
                  ... 
GKHandling           0
GKKicking            0
GKPositioning        0
GKReflexes           0
Release Clause    1275
Length: 77, dtype: int64

## Conversión de datos

En este apartado se realizara todas las conversiones necesarias en los datos que nos generan problemas, por ejemplo las columnas con valores monetarios, ya que es necesario quitar los simbolos que contengan estos datos. También pasa algo similar con ``Weight`` y ``Height``. 

Primero realizamos la función para convertir un valor monetario a un float.



In [1559]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    if not (x != x): 
    
        x = x.replace('€', '')
        ret_val = 0.0

        if type(x) == float or type(x) == int:
            ret_val = x
        if 'K' in x:
            if len(x) > 1:
                ret_val = float(x.replace('K', ''))
            ret_val = ret_val *1000
        if 'M' in x:
            if len(x) > 1:
                ret_val = float(x.replace('M', ''))
            ret_val = ret_val * 1000000.0
        return ret_val
    
    else:
        return np.nan

Después de realizar esta conversión podemos darle el formato necesario a ``Release Clause`` para poder realizar la media que antes no nos era posible. Primero aplicamos esta función a la columna de ``Release Clause`` para posteriormente sustituir los NaNs por la media de los demás valores.

In [1560]:
df["Release Clause"] = df["Release Clause"].apply(value_to_float)

In [1561]:
media = df["Release Clause"].mean()
df["Release Clause"].fillna(media, inplace = True) 
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club    Value   Wage  Special Preferred Foot  \
0         FC Barcelona  €110.5M  €565K     2202           Left   
1             Juventus     €77M  €405K     2228          Right   
2  Paris Saint-Germain  €118.5M  €290K     2143          Right   
3    Manchester United     €72M  €260K     1471          Right   
4      Manchester City    €102M  €355K     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position Height  Weight  \
0                       5.0        4.0          4.0       RF    5'7  159lbs   
1                       5.0        4.0          5.0       ST    6'2  183lbs   
2                       5.0        5.0          5.0       LW    5'9  150lbs   
3                       4.0        3.0          1.0       GK    6'4  168lbs   
4                       4.0        5.0          4.0      RCM   5'11  154lbs   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Mark

Ahora podemos observar que ya no hay NaNs en ``Release Clause``.

In [1562]:
df.isnull().sum()

Name              0
Age               0
Nationality       0
Overall           0
Potential         0
                 ..
GKHandling        0
GKKicking         0
GKPositioning     0
GKReflexes        0
Release Clause    0
Length: 77, dtype: int64

Para los datos de ``Value`` y ``Wage`` también es necesario aplicar la función **value to float** ya que también contienen símbolos que no queremos.

In [1563]:
df["Value"] = df["Value"].apply(value_to_float)
df["Wage"] = df["Wage"].apply(value_to_float)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  Special Preferred Foot  \
0         FC Barcelona  110500000.0  565000.0     2202           Left   
1             Juventus   77000000.0  405000.0     2228          Right   
2  Paris Saint-Germain  118500000.0  290000.0     2143          Right   
3    Manchester United   72000000.0  260000.0     1471          Right   
4      Manchester City  102000000.0  355000.0     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position Height  Weight  \
0                       5.0        4.0          4.0       RF    5'7  159lbs   
1                       5.0        4.0          5.0       ST    6'2  183lbs   
2                       5.0        5.0          5.0       LW    5'9  150lbs   
3                       4.0        3.0          1.0       GK    6'4  168lbs   
4                       4.0        5.0          4.0      RCM   5'11  154lbs   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0

A continuación definimos la función que utilizaremos para cambiar a un formato correcto la **altura**.

In [1564]:
def height_to_float(h):
    
    height = float(h.replace("'", "."))
    return height

In [1565]:
df["Height"] = df["Height"].apply(height_to_float)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  Special Preferred Foot  \
0         FC Barcelona  110500000.0  565000.0     2202           Left   
1             Juventus   77000000.0  405000.0     2228          Right   
2  Paris Saint-Germain  118500000.0  290000.0     2143          Right   
3    Manchester United   72000000.0  260000.0     1471          Right   
4      Manchester City  102000000.0  355000.0     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position  Height  Weight  \
0                       5.0        4.0          4.0       RF    5.70  159lbs   
1                       5.0        4.0          5.0       ST    6.20  183lbs   
2                       5.0        5.0          5.0       LW    5.90  150lbs   
3                       4.0        3.0          1.0       GK    6.40  168lbs   
4                       4.0        5.0          4.0      RCM    5.11  154lbs   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0       

También definimos la función que utilizaremos para cambiar de formato el **peso**.

In [1566]:
def weight_to_float(w):
    
    weight = float(w.replace("lbs", ""))
    return weight

In [1567]:
df["Weight"] = df["Weight"].apply(weight_to_float)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  Special Preferred Foot  \
0         FC Barcelona  110500000.0  565000.0     2202           Left   
1             Juventus   77000000.0  405000.0     2228          Right   
2  Paris Saint-Germain  118500000.0  290000.0     2143          Right   
3    Manchester United   72000000.0  260000.0     1471          Right   
4      Manchester City  102000000.0  355000.0     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position  Height  Weight  \
0                       5.0        4.0          4.0       RF    5.70   159.0   
1                       5.0        4.0          5.0       ST    6.20   183.0   
2                       5.0        5.0          5.0       LW    5.90   150.0   
3                       4.0        3.0          1.0       GK    6.40   168.0   
4                       4.0        5.0          4.0      RCM    5.11   154.0   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0       

Se puede observar que la valoración de cada jugador por posición presenta un formato incorrecto ya que tiene el símbolo "+". Por lo tanto se crea una función que elimina ese símbolo y hace la suma que está indicada. Si hay un valor NaN se devolvera ese mismo NaN ya que a continuación se tratará.

Hay una gran cantidad de jugadores, especificamente los **porteros**, que tienen demasiados NaNs, lo que hace que no sea eficiente eliminarlos porque se perdería mucha información. Entonces se hará lo mismo que con ``Release Clause``, se intercambiarán los NaNs por la **media** de esos atributos. 

In [1568]:
def new_pos_valoration(v):
    
    if not(v != v):
        val = v.split('+')
        res = float(val[0]) + float(val[1])
        return res
    else: 
        return np.nan

In [1569]:
list = ["LS","ST","RS","LW","LF","CF","RF","RW","LAM","CAM","RAM","LM","LCM","CM","RCM","RM","LWB","LDM","CDM","RDM","RWB","LB","LCB","CB","RCB","RB"]
for i in list:
    df[i] = df[i].apply(new_pos_valoration)
    df[i].fillna(df[i].mean(), inplace=True)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  Special Preferred Foot  \
0         FC Barcelona  110500000.0  565000.0     2202           Left   
1             Juventus   77000000.0  405000.0     2228          Right   
2  Paris Saint-Germain  118500000.0  290000.0     2143          Right   
3    Manchester United   72000000.0  260000.0     1471          Right   
4      Manchester City  102000000.0  355000.0     2281          Right   

   International Reputation  Weak Foot  Skill Moves Position  Height  Weight  \
0                       5.0        4.0          4.0       RF    5.70   159.0   
1                       5.0        4.0          5.0       ST    6.20   183.0   
2                       5.0        5.0          5.0       LW    5.90   150.0   
3                       4.0        3.0          1.0       GK    6.40   168.0   
4                       4.0        5.0          4.0      RCM    5.11   154.0   

          LS         ST         RS         LW         LF         CF  \
0  90.000000  90.000000  90.000000  94.000000  95.000000  95.000000   
1  94.000000  94.000000  94.000000  92.000000  93.000000  93.000000   
2  87.000000  87.000000  87.000000  92.000000  92.000000  92.000000   
3  59.842647  59.842647  59.842647  61.066809  60.750471  60.750471   
4  85.000000  85.000000  85.000000  90.000000  90.000000  90.000000   

          RF         RW        LAM        CAM        RAM         LM  \
0  95.000000  94.000000  95.000000  95.000000  95.000000  93.000000   
1  93.000000  92.000000  91.000000  91.000000  91.000000  91.000000   
2  92.000000  92.000000  92.000000  92.000000  92.000000  91.000000   
3  60.750471  61.066809  61.015195  61.015195  61.015195  61.760455   
4  90.000000  90.000000  91.000000  91.000000  91.000000  91.000000   

         LCM         CM        RCM         RM        LWB        LDM  \
0  86.000000  86.000000  86.000000  93.000000  66.000000  63.000000   
1  84.000000  84.000000  84.000000  91.000000  68.000000  64.000000   
2  84.000000  84.000000  84.000000  91.000000  68.000000  63.000000   
3  60.244569  60.244569  60.244569  61.760455  59.516891  58.831659   
4  90.000000  90.000000  90.000000  91.000000  80.000000  80.000000   

         CDM        RDM        RWB         LB        LCB         CB  \
0  63.000000  63.000000  66.000000  61.000000  49.000000  49.000000   
1  64.000000  64.000000  68.000000  64.000000  56.000000  56.000000   
2  63.000000  63.000000  68.000000  63.000000  50.000000  50.000000   
3  58.831659  58.831659  59.516891  58.806229  57.614027  57.614027   
4  80.000000  80.000000  80.000000  76.000000  69.000000  69.000000   

         RCB         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  49.000000  61.000000      84.0       95.0             70.0          90.0   
1  56.000000  64.000000      84.0       94.0             89.0          81.0   
2  50.000000  63.000000      79.0       87.0             62.0          84.0   
3  57.614027  58.806229      17.0       13.0             21.0          50.0   
4  69.000000  76.000000      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       

Para finalizar la conversión de los datos, tenemos una serie de columnas con valores no numéricos a las que tenemos que aplicar un método de conversión para que el algoritmo de entrenamiento sea capaz de leer de forma correcta estos datos. Se va a optar por substituir estas columnas por cada uno de los valores que esa variable puede tomar. Por ejemplo, en el caso de ``Nationality``, se añadiran todas las nacionalidades como columnas y si un determinado jugador es de una nacionalidad tendrá el valor 1 y si no lo es tendrá un 0. Se hará lo mismo con ``Club``, ``Prefered Foot`` y ``Position``.

In [1570]:
nt = df.pop("Nationality")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(nt, prefix='nat').reset_index(drop=True)], axis=1, sort=False)

clb = df.pop("Club")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)

prft = df.pop("Preferred Foot")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(prft, prefix='foot').reset_index(drop=True)], axis=1, sort=False)

pst = df.pop("Position")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pst, prefix='pos').reset_index(drop=True)], axis=1, sort=False)

df.head()

Name  Age  Overall  Potential        Value      Wage  Special  \
0           L. Messi   31       94         94  110500000.0  565000.0     2202   
1  Cristiano Ronaldo   33       94         94   77000000.0  405000.0     2228   
2          Neymar Jr   26       92         93  118500000.0  290000.0     2143   
3             De Gea   27       91         93   72000000.0  260000.0     1471   
4       K. De Bruyne   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight  \
0                       5.0        4.0          4.0    5.70   159.0   
1                       5.0        4.0          5.0    6.20   183.0   
2                       5.0        5.0          5.0    5.90   150.0   
3                       4.0        3.0          1.0    6.40   168.0   
4                       4.0        5.0          4.0    5.11   154.0   

          LS         ST         RS         LW         LF         CF  \
0  90.000000  90.000000  90.000000  94.000000  95.000000  95.000000   
1  94.000000  94.000000  94.000000  92.000000  93.000000  93.000000   
2  87.000000  87.000000  87.000000  92.000000  92.000000  92.000000   
3  59.842647  59.842647  59.842647  61.066809  60.750471  60.750471   
4  85.000000  85.000000  85.000000  90.000000  90.000000  90.000000   

          RF         RW        LAM        CAM        RAM         LM  \
0  95.000000  94.000000  95.000000  95.000000  95.000000  93.000000   
1  93.000000  92.000000  91.000000  91.000000  91.000000  91.000000   
2  92.000000  92.000000  92.000000  92.000000  92.000000  91.000000   
3  60.750471  61.066809  61.015195  61.015195  61.015195  61.760455   
4  90.000000  90.000000  91.000000  91.000000  91.000000  91.000000   

         LCM         CM        RCM         RM        LWB        LDM  \
0  86.000000  86.000000  86.000000  93.000000  66.000000  63.000000   
1  84.000000  84.000000  84.000000  91.000000  68.000000  64.000000   
2  84.000000  84.000000  84.000000  91.000000  68.000000  63.000000   
3  60.244569  60.244569  60.244569  61.760455  59.516891  58.831659   
4  90.000000  90.000000  90.000000  91.000000  80.000000  80.000000   

         CDM        RDM        RWB         LB        LCB         CB  \
0  63.000000  63.000000  66.000000  61.000000  49.000000  49.000000   
1  64.000000  64.000000  68.000000  64.000000  56.000000  56.000000   
2  63.000000  63.000000  68.000000  63.000000  50.000000  50.000000   
3  58.831659  58.831659  59.516891  58.806229  57.614027  57.614027   
4  80.000000  80.000000  80.000000  76.000000  69.000000  69.000000   

         RCB         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  49.000000  61.000000      84.0       95.0             70.0          90.0   
1  56.000000  64.000000      84.0       94.0             89.0          81.0   
2  50.000000  63.000000      79.0       87.0             62.0          84.0   
3  57.614027  58.806229      17.0       13.0             21.0          50.0   
4  69.000000  76.000000      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina 

Para finalizar, se elimina la columna ``Name`` como se había indicado al principio de la práctica.

In [1571]:
df.drop(['Name'],axis=1,inplace=True)
df.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight  \
0                       5.0        4.0          4.0    5.70   159.0   
1                       5.0        4.0          5.0    6.20   183.0   
2                       5.0        5.0          5.0    5.90   150.0   
3                       4.0        3.0          1.0    6.40   168.0   
4                       4.0        5.0          4.0    5.11   154.0   

          LS         ST         RS         LW         LF         CF  \
0  90.000000  90.000000  90.000000  94.000000  95.000000  95.000000   
1  94.000000  94.000000  94.000000  92.000000  93.000000  93.000000   
2  87.000000  87.000000  87.000000  92.000000  92.000000  92.000000   
3  59.842647  59.842647  59.842647  61.066809  60.750471  60.750471   
4  85.000000  85.000000  85.000000  90.000000  90.000000  90.000000   

          RF         RW        LAM        CAM        RAM         LM  \
0  95.000000  94.000000  95.000000  95.000000  95.000000  93.000000   
1  93.000000  92.000000  91.000000  91.000000  91.000000  91.000000   
2  92.000000  92.000000  92.000000  92.000000  92.000000  91.000000   
3  60.750471  61.066809  61.015195  61.015195  61.015195  61.760455   
4  90.000000  90.000000  91.000000  91.000000  91.000000  91.000000   

         LCM         CM        RCM         RM        LWB        LDM  \
0  86.000000  86.000000  86.000000  93.000000  66.000000  63.000000   
1  84.000000  84.000000  84.000000  91.000000  68.000000  64.000000   
2  84.000000  84.000000  84.000000  91.000000  68.000000  63.000000   
3  60.244569  60.244569  60.244569  61.760455  59.516891  58.831659   
4  90.000000  90.000000  90.000000  91.000000  80.000000  80.000000   

         CDM        RDM        RWB         LB        LCB         CB  \
0  63.000000  63.000000  66.000000  61.000000  49.000000  49.000000   
1  64.000000  64.000000  68.000000  64.000000  56.000000  56.000000   
2  63.000000  63.000000  68.000000  63.000000  50.000000  50.000000   
3  58.831659  58.831659  59.516891  58.806229  57.614027  57.614027   
4  80.000000  80.000000  80.000000  76.000000  69.000000  69.000000   

         RCB         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  49.000000  61.000000      84.0       95.0             70.0          90.0   
1  56.000000  64.000000      84.0       94.0             89.0          81.0   
2  50.000000  63.000000      79.0       87.0             62.0          84.0   
3  57.614027  58.806229      17.0       13.0             21.0          50.0   
4  69.000000  76.000000      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0      

Comprobamos que no queda nigun valor nulo en el dataset.

In [1572]:
df.isnull().sum().sum()

0

## Predicción

Ahora que tenemos nuestro dataset listo para el entrenamiento, se saca la columna ``Value`` ya que será nuestra variable a predecir. Se realizará el entrenamiento con el 20% de nuestro dataset.

In [1573]:
val = df.pop("Value")

In [1574]:
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.2, random_state=50)
len(X_train)

14334

Ahora entrenaremos un modelo de regresión lineal.

In [1575]:
reg = linear_model.LinearRegression().fit(X_train, y_train) 

Finalmente obtenemos una métrica 𝑅2 para la regresión, usamos la implementación de scikit-learn.

In [1576]:
preds = reg.predict(X_test)


In [1577]:
y_test

4020     2700000.0
9164      475000.0
3153     2700000.0
7790      725000.0
9097      250000.0
           ...    
9696      800000.0
245     24500000.0
8039      750000.0
9708     1300000.0
9066      600000.0
Name: Value, Length: 3584, dtype: float64

In [1578]:
preds[0]

3098005.766248701

In [1579]:
r2_score(preds, y_test)

0.939901711639472

Para concluir, se puede analizar los resultados obtenidos fijandose en el valor predicho (3097980.37), el cual se acerca significamente al valor real (2700000). La predicción tiene aproxidamente un 94% de acierto así que se puede decir que es un buen resultado para el entrenamiento.